### PR TITLE
Added Pull request for development branch

### PR DESCRIPTION
**Pull request by vinaybajjuri58**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+app.use("/api/categories", categoryRouter);
+app.use("/api/videos", videoRouter);
+app.use("/api/playlists", playlistRouter);
+
+// error Handler
+app.use(errorHandler);
+
+// Do not move pathnotfound handler
+app.use(pathNotFoundHandler);
+initialiseDBConnection();
```
Comment (Aravindan07) [inline index.js]: It is better to initialize the DB connection before every routes i.e., after line 14.

File: db/db.connect.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
+const dbUrl = process.env.DB_URL;
```
Comment (Aravindan07) [inline db/db.connect.js]: Initializing an unnecessary variable here you can directly pass the DB_URL in the arguments.

File: models/User.model.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+      type: Schema.Types.ObjectId,
+      ref: "Playlist",
+    },
+  ],
+});
+const User = mongoose.model("User", userSchema);
+module.exports = {
+  User,
+};
```
Comment (Aravindan07) [inline models/User.model.js]: These lines can be reduced into a single line like ``` module.exports = mongoose.model("User",UserSchema); ```

File: controllers/playlist.controller.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      playlist,
+    });
+  }
+};
+const getPlaylistVideos = async (req, res) => {
+  const playlistId = req.playlist._id;
+  try {
+    const playlistVideosData = await Playlist.findById(playlistId)
+      .populate("videos")
+      .select("__v");
```
Comment (Aravindan07) [inline controllers/playlist.controller.js]: If you want to remove __v from data then add a - sign before that e.g., `.select("-__v")`

File: controllers/video.controller.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+const getAllVideos = async (req, res) => {
+  try {
+    const videos = await Video.find({});
+    res.status(200).json({
+      success: true,
+      message: "All video data",
+      videos,
+    });
+  } catch (err) {
+    res.status(400).json({
```
Comment (Aravindan07) [inline controllers/video.controller.js]: This is a 500 error I hope. If it is so please change it.

File: controllers/category.controller.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+const getAllCategories = async (req, res) => {
+  try {
+    const allCategories = await Category.find({});
+    res.status(200).json({
+      success: true,
+      message: "All Categories",
+      categories: allCategories,
+    });
+  } catch (err) {
+    res.status(400).json({
```
Comment (Aravindan07) [inline controllers/category.controller.js:14]: Here also it's 500 I think. if not then it is fine.

File: controllers/category.controller.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      success: false,
+      message: "Error in getting categories",
+      errMessage: err.errMessage,
+    });
+  }
+};
+const addACategory = async (req, res) => {
+  const categoryDetails = req.body;
+  const newCategory = new Category(categoryDetails);
+  let savedCategory;
```
Comment (Aravindan07) [inline controllers/category.controller.js]: Declare it as const you are not changing its value here, right? Then change from let to const

**Review by Aravindan07 — CHANGES_REQUESTED** [review]
Great Work Vinay your code looks clean and readable.

File: controllers/category.controller.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    });
+  }
+};
+const updateCategory = async (req, res) => {
+  let category = req.category;
+  const updateCategory = req.body;
+  category = extend(category, updateCategory);
+  try {
+    const updatedCategory = await category.save();
+    res.status(204).json({
```
Comment (Jaynil1611) [inline controllers/category.controller.js]: Here the status should be 201 for `post` update operation
```suggestion
    res.status(201).json({
```

File: controllers/category.controller.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    });
+  } catch (err) {
+    res.status(400).json({
+      success: false,
+      message: "Error in getting categories",
+      errMessage: err.errMessage,
+    });
+  }
+};
+const addACategory = async (req, res) => {
```
Comment (Jaynil1611) [inline controllers/category.controller.js:22]: Naming of this function could be better like `postCategory` or only `addCategory`
Similarly, everywhere you have used `A` for singular operation, you can rename all such functions

File: controllers/category.controller.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      category: savedCategory,
+    });
+  } catch (err) {
+    res.status(500).json({
+      success: false,
+      message: "Error in adding category",
+      errMessage: err,
+    });
+  }
+};
```
Comment (Jaynil1611) [inline controllers/category.controller.js:39]: Leave a line after each function completion

File: controllers/category.controller.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    message: "Category details",
+    category: req.category,
+  });
+};
+const addAVideo = async (req, res) => {
+  const category = req.category;
+  const videoDetails = { ...req.body, category: category._id };
+  const newVideo = new Video(videoDetails);
+  try {
+    const sess = await mongoose.startSession();
```
Comment (Jaynil1611) [inline controllers/category.controller.js]: Use meaningful variable naming `sess` to `session`

File: controllers/paramHandlers.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+const { Category } = require("../models/Category.model");
+const { Video } = require("../models/Video.model");
+const { Playlist } = require("../models/Playlist.model");
+const categoryParamHandler = async (req, res, next, categoryId) => {
+  try {
+    const category = await Category.findById(categoryId).populate("videos");
```
Comment (Jaynil1611) [inline controllers/paramHandlers.js]: Here you should have another `try catch` block if the category is not found, then you also should return `res.status(400) ` with `Not found` message

File: controllers/paramHandlers.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      success: false,
+      message: "Error in getting category Details",
+      errMessage: err.errMessage,
+    });
+  }
+};
+const playlistParamHandler = async (req, res, next, playlistId) => {
+  try {
+    const playlistData = await Playlist.findById(playlistId);
+    if (playlistData === null) {
```
Comment (Jaynil1611) [inline controllers/paramHandlers.js]: This comparison can be simplified by
```suggestion
    if (playlistData) {
```

File: controllers/playlist.controller.js [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  } catch (err) {
+    req.status(400).json({
+      success: false,
+      message: "Error in getting playlist videos",
+      errMessage: err.errMessage,
+    });
+  }
+};
+const addVideoToPlaylist = async (req, res) => {
+  const videoId = req.body.video;
```
Comment (Jaynil1611) [inline controllers/playlist.controller.js]: This should be `req.body.videoId`

**Review by Jaynil1611 — CHANGES_REQUESTED** [review]
Great work @vinaybajjuri58 
I have added few suggestions:

**Review by Jaynil1611 — APPROVED** [review]
I am approving the changes

**Review by Aravindan07 — APPROVED** [review]

**Comment by Aravindan07** [comment]
You have incorporated all the changes which I asked great work!

**Pull request merged**